### ISSUE TITLE
Inconsistent Ticker Data for Bybit linear futures via websockets

### ISSUE THREAD

**Issue by scoriiu** [ccxt/ccxt#22785]
### Operating System

MacOS

### Programming Languages

Python

### CCXT Version

4.2.55

### Description

The returned ticker for linear futures (perp) does not contain all the necessary information (e.g., fundingRate) sometimes. This issue arises from incorrect determination of the isSpot variable within the ticker function.

Currently, isSpot is set as follows:
`isSpot = self.safe_string(data, 'fundingRate') is None`
This approach is flawed because it relies on the presence of the fundingRate field, which may not be included in delta messages. As a result, isSpot can be incorrectly set to True for perpetual contract updates, causing missing information.

When isSpot is explicitly set to False, the function works correctly. Therefore, I propose that we determine whether the market is a spot or perpetual contract based on the market data (e.g. self.markets) instead of relying on the presence of specific fields in the data.

`    def handle_ticker(self, client: Client, message):
        #
        # linear
        #     {
        #         "topic": "tickers.BTCUSDT",
        #         "type": "snapshot",
        #         "data": {
        #             "symbol": "BTCUSDT",
        #             "tickDirection": "PlusTick",
        #             "price24hPcnt": "0.017103",
        #             "lastPrice": "17216.00",
        #             "prevPrice24h": "16926.50",
        #             "highPrice24h": "17281.50",
        #             "lowPrice24h": "16915.00",
        #             "prevPrice1h": "17238.00",
        #             "markPrice": "17217.33",
        #             "indexPrice": "17227.36",
        #             "openInterest": "68744.761",
        #             "openInterestValue": "1183601235.91",
        #             "turnover24h": "1570383121.943499",
        #             "volume24h": "91705.276",
        #             "nextFundingTime": "1673280000000",
        #             "fundingRate": "-0.000212",
        #             "bid1Price": "17215.50",
        #             "bid1Size": "84.489",
        #             "ask1Price": "17216.00",
        #             "ask1Size": "83.020"
        #         },
        #         "cs": 24987956059,
        #         "ts": 1673272861686
        #     }
        #
        # option
        #     {
        #         "id": "tickers.BTC-6JAN23-17500-C-2480334983-1672917511074",
        #         "topic": "tickers.BTC-6JAN23-17500-C",
        #         "ts": 1672917511074,
        #         "data": {
        #             "symbol": "BTC-6JAN23-17500-C",
        #             "bidPrice": "0",
        #             "bidSize": "0",
        #             "bidIv": "0",
        #             "askPrice": "10",
        #             "askSize": "5.1",
        #             "askIv": "0.514",
        #             "lastPrice": "10",
        #             "highPrice24h": "25",
        #             "lowPrice24h": "5",
        #             "markPrice": "7.86976724",
        #             "indexPrice": "16823.73",
        #             "markPriceIv": "0.4896",
        #             "underlyingPrice": "16815.1",
        #             "openInterest": "49.85",
        #             "turnover24h": "446802.8473",
        #             "volume24h": "26.55",
        #             "totalVolume": "86",
        #             "totalTurnover": "1437431",
        #             "delta": "0.047831",
        #             "gamma": "0.00021453",
        #             "vega": "0.81351067",
        #             "theta": "-19.9115368",
        #             "predictedDeliveryPrice": "0",
        #             "change24h": "-0.33333334"
        #         },
        #         "type": "snapshot"
        #     }
        #
        # spot
        #     {
        #         "topic": "tickers.BTCUSDT",
        #         "ts": 1673853746003,
        #         "type": "snapshot",
        #         "cs": 2588407389,
        #         "data": {
        #             "symbol": "BTCUSDT",
        #             "lastPrice": "21109.77",
        #             "highPrice24h": "21426.99",
        #             "lowPrice24h": "20575",
        #             "prevPrice24h": "20704.93",
        #             "volume24h": "6780.866843",
        #             "turnover24h": "141946527.22907118",
        #             "price24hPcnt": "0.0196",
        #             "usdIndexPrice": "21120.2400136"
        #         }
        #     }
        #
        # lt ticker
        #     {
        #         "topic": "tickers_lt.EOS3LUSDT",
        #         "ts": 1672325446847,
        #         "type": "snapshot",
        #         "data": {
        #             "symbol": "EOS3LUSDT",
        #             "lastPrice": "0.41477848043290448",
        #             "highPrice24h": "0.435285472510871305",
        #             "lowPrice24h": "0.394601507960931382",
        #             "prevPrice24h": "0.431502290172376349",
        #             "price24hPcnt": "-0.0388"
        #         }
        #     }
        #
     
        topic = self.safe_string(message, 'topic', '')
        updateType = self.safe_string(message, 'type', '')
        data = self.safe_value(message, 'data', {})
        isSpot = self.safe_string(data, 'fundingRate') is None # <<<<<<<<<< Causes the issue
        type = 'spot' if isSpot else 'contract'
        symbol = None
        parsed = None
        if (updateType == 'snapshot') or isSpot:
            parsed = self.parse_ticker(data)
            symbol = parsed['symbol']
        elif updateType == 'delta':
            topicParts = topic.split('.')
            topicLength = len(topicParts)
            marketId = self.safe_string(topicParts, topicLength - 1)
            market = self.safe_market(marketId, None, None, type)
            symbol = market['symbol']
            # update the info in place
            ticker = self.safe_value(self.tickers, symbol, {})
            rawTicker = self.safe_value(ticker, 'info', {})
            merged = self.extend(rawTicker, data)
            parsed = self.parse_ticker(merged)
        timestamp = self.safe_integer(message, 'ts')
        parsed['timestamp'] = timestamp
        parsed['datetime'] = self.iso8601(timestamp)
        self.tickers[symbol] = parsed
        messageHash = 'ticker:' + symbol
        client.resolve(self.tickers[symbol], messageHash)`

### Code

```
      async def watch_tickers(self):
        while True:
            try:
                tickers = await self.exchange.ccxt.watch_tickers(
                    symbols=[
                        'SSV/USDT:USDT', 'DOT/USDT:USDT', 'GAL/USDT:USDT', 'KLAY/USDT:USDT', 
                        'CAKE/USDT:USDT', 'STRK/USDT:USDT', 'RUNE/USDT:USDT', 'SUI/USDT:USDT', 
                        'MAGIC/USDT:USDT', 'FLOW/USDT:USDT'
                    ]
                )
                # self.logger.info(tickers)
                for symbol, ticker in tickers.items():
                    if symbol not in self.hasFunding:
                        if ticker["info"].get("fundingRate"):
                            self.hasFunding[symbol] = True
                    self._update_ticker_data(symbol, ticker)

                # self.logger.info(f"Tickers {len(list(self.tickers.values()))}")
            except asyncio.CancelledError:
                break
            except Exception as e:
                print(f"Error watching order: {repr(e)}")
                await asyncio.sleep(5)

```

**Comment by carlosmiei** [comment]
Hello @scoriiu,
Thanks for reporting it. Quick question, have you tried to update your ccxt version to the latest to see if the issue persists?


> y. Therefore, I propose that we determine whether the market is a spot or perpetual contract based on the market data (e.g. self.markets) instead of relying on the presence of specific fields in the data.

Regarding this, it might be tricky because bybit has conflicting market ids between spot and linear swaps, nonetheless, we will try to find a solution.

**Comment by carlosmiei** [comment]
@scoriiu I just confirmed that we're now using a different check to infer if the market is spot or not, so I believe this issue is resolved in the latest version. Can you please update and try again?

![image](https://github.com/ccxt/ccxt/assets/43336371/379b44d2-a499-490a-be12-93628149798b)

**Comment by scoriiu** [comment]
@carlosmiei Upgraded to `4.3.42` and it seems like the new check is doing the job.  Thanks, the issue can be closed.

**Comment by scoriiu** [comment]
Fixed in `4.3.2`

**Comment by carlosmiei** [comment]
@scoriiu No problem, let us know if anything!
